### DISCUSSION TITLE
Sokol drawing limited to 16384 instructions

### DISCUSSION THREAD

**Discussion by benstigsen** [vlang/v#11897] [Ideas]
**What did you do?**
```v
for x in 0..150 {
  for y in 0..110 {
    app.gg.set_pixel(x, y, gx.white)
  }
}
```

**What did you expect to see?**
A white rectangle
 
**What did you see instead?**
Nothing

**Description**
Due to the sokol_gl.h header, it is only possible to give it a max of `1 << 14` or `16384` commands. Everytime `set_pixel()` is called, that counts as a command. While this isn't unintended behavior, it can be pretty limiting. Finding ways around the restrictions is pretty easy by drawing and updating an image instead, but for some cases I might prefer to draw pixel by pixel.
I'd request for these limits to be increased: https://github.com/vlang/v/blob/master/thirdparty/sokol/util/sokol_gl.h#L2031-L2032

_I wasn't sure whether to mark this as a feature request or bug_

**Comment by JalonSolov** [comment]
Feature request since it is working as designed, now.

**Comment by danieldaeschle** [comment]
Is it even related to V?

**Comment by medvednikov** [comment]
Of course, it's our graphical backend.

@BenStigsen can you please  also report it on https://github.com/floooh/sokol ?

**Comment by benstigsen** [comment]
Sure!

**Comment by benstigsen** [comment]
https://github.com/floooh/sokol/issues/540

**Comment by medvednikov** [comment]
According to the author:

> the limits you're seeing are just the default values, you can increase those limits during initialization:

**Comment by benstigsen** [comment]
Yep, that means though, that either one should create their own function initializing it, or one should have the possibility to modify the `C.sgl_desc_t{}` before initialization. https://github.com/vlang/v/blob/master/vlib/gg/gg.v#L173-L174

**Comment by medvednikov** [comment]
Yeah that has to be added to `gg`.

**Comment by benstigsen** [comment]
~~After looking into it further, I think the drawing is being done with `gg` in a weird way.~~
~~https://github.com/vlang/v/blob/master/thirdparty/sokol/util/sokol_gl.h#L559~~
> ~~each begin/end is one draw command~~

~~This means that if I draw 5 rectangles, this counts as 5 draw commands, even though it will all be shown at the same time any way. Surely some optimization should be possible, that doesn't require each shape drawn to be surrounded in `sgl.begin_xxx()` and `sgl.end()`. Then instead make `app.ctx.begin()` call `sgl.begin_points()` and make `app.ctx.end()` call `sgl.end()`,~~

**Comment by benstigsen** [comment]
**Scratch what I said above**
> each begin/end is one draw command

This does not seem to be true unless V changed the commands set to Sokol.

I tested this with my own set_pixel() function
```v
pub fn set_pixel(mut ctx &gg.Context, x f32, y f32, c gx.Color) {
  sgl.v2f(x * ctx.scale, y * ctx.scale)
  sgl.v2f((x + 1) * ctx.scale, y * ctx.scale)
  sgl.v2f((x + 1) * ctx.scale, (y + 1) * ctx.scale)
  sgl.v2f(x * ctx.scale, (y + 1) * ctx.scale)
}

pub fn test(mut ctx &gg.Context) {
  c := gx.white

  if c.a != 255 {
    sgl.load_pipeline(ctx.timage_pip)
  }

  sgl.c4b(c.r, c.g, c.b, c.a)
  sgl.begin_quads()
  for x in 0..150 {
    for y in 0..110 {
      set_pixel(mut ctx, x, y, c)
    }
  }
  sgl.end()
}
```
The example above still doesn't draw even though we're only calling `sgl.being_quads()` and `sgl.end()` once.

It depends purely on the amount of commands, so the max should be increased, preferably to something ***big*** so that it could support filling something like an entire screen (1280x720, 1920x1080, etc...)

**Comment by JalonSolov** [comment]
3840x2160... :-)

**Comment by benstigsen** [comment]
Creating a static variable, and keeping track of the amount of pixels drawn, is useless since force redrawing can lead to cut off shapes and other stuff, which makes `sgl.draw()` and `ctx.end()` a bad solution.

Does seem like the only solution is to increase the max commands and vertices.

A separate thing, I'd also propose an option to have a more performant `set_pixel()` built in to `gg`. Currently a pixel is drawn like this:
```v
if c.a != 255 {
  sgl.load_pipeline(ctx.timage_pip)
}
sgl.c4b(c.r, c.g, c.b, c.a)
sgl.begin_quads()
// pixel drawing
sgl.end()
```
This means that if I were to draw something like 1000x1000 pixels, it would call **EACH** of the functions above 1,000,000 times, for one draw.

**Comment by medvednikov** [comment]
for sure, can have something like `set_pixels()`

**Comment by benstigsen** [comment]
After looking at it more, the max amount of commands shouldn't be set to something specific. Each command is 24 bytes, so if we were to set max draw commands to something like:
`3840x2160 = 8,294,400`, that would mean we're allocating `199,065,600 bytes = ~200 MB` each frame.

https://github.com/vlang/v/blob/master/thirdparty/sokol/util/sokol_gl.h#L368
https://github.com/vlang/v/blob/master/thirdparty/sokol/util/sokol_gl.h#L2659

Instead, `gg` should expose `C.sgl_desc_t{}` where we can set it ourselves, and modify it frame by frame if wanted.
and/or
There could be kept track of how many pixels are being drawn, if we reach the limit, bit shift it by 1 and reallocate _(this would however cause trouble if someone were to draw with Sokol directly, since the value is not being kept track of anymore)._

**Comment by benstigsen** [comment]
> for sure, can have something like `set_pixels()`

I could work on this, but do you think the function should be marked as `[unsafe]`? I think this would be best since it requires the user to load the pipeline, begin quads/line_strip and so on themselves. As soon as the developer forgets one of the 4 function calls, it can cause an error with Sokol.
```v
if c.a != 255 {
  sgl.load_pipeline(ctx.timage_pip)
}
sgl.c4b(c.r, c.g, c.b, c.a)
sgl.begin_quads()
gg.set_pixels([x0, y0, x1, y1, etc..])
sgl.end()
```

**Comment by medvednikov** [comment]
> Instead, gg should expose C.sgl_desc_t{} where we can set it ourselves, and modify it frame by frame if wanted.

yes, definitely

@BenStigsen ok, unsafe sounds good

**Comment by benstigsen** [comment]
> A separate thing, I'd also propose an option to have a more performant `set_pixel()` built in to `gg`. Currently a pixel is drawn like this:
> ```v
> if c.a != 255 {
>   sgl.load_pipeline(ctx.timage_pip)
> }
> sgl.c4b(c.r, c.g, c.b, c.a)
> sgl.begin_quads()
> // pixel drawing
> sgl.end()
> ```
> This means that if I were to draw something like 1000x1000 pixels, it would call **EACH** of the functions above 1,000,000 times, for one draw.

`set_pixels()` https://github.com/vlang/v/pull/11236

**Comment by benstigsen** [comment]
> 3840x2160... :-)

I just realized, that since each command is 24 bytes, that would mean that filling something like a screen with `3840x2160` pixels (using `set_pixel()`) is going to take 200 MB of memory. Not sure whether to consider this a lot or not. Of course, filling a screen with `set_pixel()` isn't ideal when doing something like `draw_filled_rect()` only takes a few bytes to do the same any way. Hmm, might be worth looking into this in the future, depending on whether that should be the user's job to optimize or `gg`'s job.
